packages/vscode-extension/src/project/metro.ts

```diff
@@ -292,6 +292,82 @@ export class Metro implements Disposable {
     ws.close();
   }
 
+  private async evaluateJsFromCdp(
+    webSocketDebuggerUrl: string,
+    source: string,
+    timeoutMs: number = 2000
+  ): Promise<string | undefined> {
+    const REQUEST_ID = 0;
+    let timeoutHandle: NodeJS.Timeout;
+
+    const wsUrl = new URL(webSocketDebuggerUrl);
+    wsUrl.hostname = "localhost";
+    wsUrl.port = String(this._port);
+
+    return new Promise((resolve, reject) => {
+      let settled = false;
+      const ws = new WebSocket(wsUrl);
+
+      timeoutHandle = setTimeout(() => {
+        Logger.debug(`[evaluateJsFromCdpAsync] Request timeout from ${wsUrl.toString()}`);
+        reject(new Error("Request timeout"));
+        settled = true;
+        ws.close();
+      }, timeoutMs);
+
+      ws.on("open", () => {
+        ws.send(
+          JSON.stringify({
+            id: REQUEST_ID,
+            method: "Runtime.evaluate",
+            params: { expression: source },
+          })
+        );
+      });
+
+      ws.on("error", (e) => {
+        Logger.debug(`[evaluateJsFromCdpAsync] Failed to connect ${wsUrl.toString()}`, e);
+        reject(e);
+        settled = true;
+        clearTimeout(timeoutHandle);
+        ws.close();
+      });
+
+      ws.on("close", () => {
+        if (!settled) {
+          reject(new Error("WebSocket closed before response was received."));
+          clearTimeout(timeoutHandle);
+        }
+      });
+
+      ws.on("message", (data) => {
+        Logger.debug(
+          `[evaluateJsFromCdpAsync] message received from ${wsUrl.toString()}: ${data.toString()}`
+        );
+        try {
+          const response = JSON.parse(data.toString());
+          if (response.id === REQUEST_ID) {
+            if (response.error) {
+              reject(new Error(response.error.message));
+            } else if (response.result.result.type === "string") {
+              resolve(response.result.result.value);
+            } else {
+              resolve(undefined);
+            }
+            settled = true;
+            clearTimeout(timeoutHandle);
+            ws.close();
```

Comment:
```
             settled = true;
             clearTimeout(timeoutHandle);
             ws.close();
 ```
 It repeated x times, let's extract it to some method

Reply:
thx nice catch